#Importation des librairies et des bibliothèques

In [ ]:
# Importer les bibliothèques
import pandas as pd
import os
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Importation du fichier et exploration du dataframe

In [ ]:
# Monter Google Drive pour accéder aux fichiers
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Chemin vers le dossier contenant le fichier CSV
dossier = '/content/drive/My Drive/Colab Notebooks/Projet_Facebook_Dataset'

In [ ]:
# Chemin vers le dossier contenant le fichier CSV
dossier = '/content/drive/My Drive/Colab Notebooks/Projet_Facebook_Dataset'
chemin_fichier = os.path.join(dossier, 'dataset_Facebook.csv')

# Lire le fichier CSV dans un DataFrame pandas en spécifiant le séparateur
df = pd.read_csv(chemin_fichier, sep=';')

# Affichage des 5 premières lignes du DataFrame
display(df.head())

Page total likes    Type  Category  Post Month  Post Weekday  Post Hour  \
0            139441   Photo         2          12             4          3   
1            139441  Status         2          12             3         10   
2            139441   Photo         3          12             3          3   
3            139441   Photo         2          12             2         10   
4            139441   Photo         2          12             2          3   

   Paid  Lifetime Post Total Reach  Lifetime Post Total Impressions  \
0     0                       2752                             5091   
1     0                      10460                            19057   
2     0                       2413                             4373   
3     1                      50128                            87991   
4     0                       7244                            13594   

   Lifetime Engaged Users  Lifetime Post Consumers  \
0                     178                      109   
1                    1457                     1361   
2                     177                      113   
3                    2211                      790   
4                     671                      410   

   Lifetime Post Consumptions  \
0                         159   
1                        1674   
2                         154   
3                        1119   
4                         580   

   Lifetime Post Impressions by people who have liked your Page  \
0                                               3078              
1                                              11710              
2                                               2812              
3                                              61027              
4                                               6228              

   Lifetime Post reach by people who like your Page  \
0                                              1640   
1                                              6112   
2                                              1503   
3                                             32048   
4                                              3200   

   Lifetime People who have liked your Page and engaged with your post  \
0                                                119                     
1                                               1108                     
2                                                132                     
3                                               1386                     
4                                                396                     

   comment    like  share  Total Interactions  
0        4    79.0   17.0                 100  
1        5   130.0   29.0                 164  
2        0    66.0   14.0                  80  
3       58  1572.0  147.0                1777  
4       19   325.0   49.0                 393

In [ ]:
print(f"Nombre de lignes : {len(df)}")

Nombre de lignes : 500


In [ ]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 19 columns):
 #   Column                                                               Non-Null Count  Dtype  
---  ------                                                               --------------  -----  
 0   Page total likes                                                     500 non-null    int64  
 1   Type                                                                 500 non-null    object 
 2   Category                                                             500 non-null    int64  
 3   Post Month                                                           500 non-null    int64  
 4   Post Weekday                                                         500 non-null    int64  
 5   Post Hour                                                            500 non-null    int64  
 6   Paid                                                                 500 non-null    int64  
 7   Lifetime

None

In [ ]:
display(df.describe())

Page total likes    Category  Post Month  Post Weekday   Post Hour  \
count        500.000000  500.000000  500.000000    500.000000  500.000000   
mean      123194.176000    1.880000    7.038000      4.150000    7.840000   
std        16272.813214    0.852675    3.307936      2.030701    4.368589   
min        81370.000000    1.000000    1.000000      1.000000    1.000000   
25%       112676.000000    1.000000    4.000000      2.000000    3.000000   
50%       129600.000000    2.000000    7.000000      4.000000    9.000000   
75%       136393.000000    3.000000   10.000000      6.000000   11.000000   
max       139441.000000    3.000000   12.000000      7.000000   23.000000   

             Paid  Lifetime Post Total Reach  Lifetime Post Total Impressions  \
count  500.000000                  500.00000                     5.000000e+02   
mean     0.278000                13903.36000                     2.958595e+04   
std      0.448462                22740.78789                     7.680325e+04   
min      0.000000                  238.00000                     5.700000e+02   
25%      0.000000                 3315.00000                     5.694750e+03   
50%      0.000000                 5281.00000                     9.051000e+03   
75%      1.000000                13168.00000                     2.208550e+04   
max      1.000000               180480.00000                     1.110282e+06   

       Lifetime Engaged Users  Lifetime Post Consumers  \
count              500.000000               500.000000   
mean               920.344000               798.772000   
std                985.016636               882.505013   
min                  9.000000                 9.000000   
25%                393.750000               332.500000   
50%                625.500000               551.500000   
75%               1062.000000               955.500000   
max              11452.000000             11328.000000   

       Lifetime Post Consumptions  \
count                  500.000000   
mean                  1415.130000   
std                   2000.594118   
min                      9.000000   
25%                    509.250000   
50%                    851.000000   
75%                   1463.000000   
max                  19779.000000   

       Lifetime Post Impressions by people who have liked your Page  \
count                                       5.000000e+02              
mean                                        1.676638e+04              
std                                         5.979102e+04              
min                                         5.670000e+02              
25%                                         3.969750e+03              
50%                                         6.255500e+03              
75%                                         1.486050e+04              
max                                         1.107833e+06              

       Lifetime Post reach by people who like your Page  \
count                                        500.000000   
mean                                        6585.488000   
std                                         7682.009405   
min                                          236.000000   
25%                                         2181.500000   
50%                                         3417.000000   
75%                                         7989.000000   
max                                        51456.000000   

       Lifetime People who have liked your Page and engaged with your post  \
count                                         500.000000                     
mean                                          609.986000                     
std                                           612.725618                     
min                                             9.000000                     
25%                                           291.000000                     
50%                                           412.000000                     
75%     

## Nettoyage des données

In [ ]:
# Est-ce qu'il y a des valeurs nulles ?
df.isnull().sum()

Page total likes                                                       0
Type                                                                   0
Category                                                               0
Post Month                                                             0
Post Weekday                                                           0
Post Hour                                                              0
Paid                                                                   0
Lifetime Post Total Reach                                              0
Lifetime Post Total Impressions                                        0
Lifetime Engaged Users                                                 0
Lifetime Post Consumers                                                0
Lifetime Post Consumptions                                             0
Lifetime Post Impressions by people who have liked your Page           0
Lifetime Post reach by people who like your Page                       0
Lifetime People who have liked your Page and engaged with your post    0
comment                                                                0
like                                                                   1
share                                                                  4
Total Interactions                                                     0
dtype: int64

In [ ]:
# Est-ce qu'il y a des doublons ?
df.duplicated().sum()

np.int64(0)

In [ ]:
# Gérer les valeurs manquantes avant de convertir en int64.
# Utilisation de la médiane pour l'imputation afin d'éviter les erreurs avec astype(int).

df['like'] = df['like'].fillna(df['like'].median())
df['share'] = df['share'].fillna(df['share'].median())

# Changer le type des colonnes 'like' et 'share' en int64
df['like'] = df['like'].astype('int64')
df['share'] = df['share'].astype('int64')

# La colonne 'Category' n'est pas pertinente car il n'y a aucune information écrite dans la description du dataset.
# Suppression de la colonne 'Category', ignorer l'erreur si elle n'existe pas.
df.drop('Category', axis=1, inplace=True, errors='ignore')

print("DataFrame après modifications:")
display(df.info())
display(df.head())

DataFrame après modifications:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 18 columns):
 #   Column                                                               Non-Null Count  Dtype 
---  ------                                                               --------------  ----- 
 0   Page total likes                                                     500 non-null    int64 
 1   Type                                                                 500 non-null    object
 2   Post Month                                                           500 non-null    int64 
 3   Post Weekday                                                         500 non-null    int64 
 4   Post Hour                                                            500 non-null    int64 
 5   Paid                                                                 500 non-null    int64 
 6   Lifetime Post Total Reach                                            500 non-null  

None

Page total likes    Type  Post Month  Post Weekday  Post Hour  Paid  \
0            139441   Photo          12             4          3     0   
1            139441  Status          12             3         10     0   
2            139441   Photo          12             3          3     0   
3            139441   Photo          12             2         10     1   
4            139441   Photo          12             2          3     0   

   Lifetime Post Total Reach  Lifetime Post Total Impressions  \
0                       2752                             5091   
1                      10460                            19057   
2                       2413                             4373   
3                      50128                            87991   
4                       7244                            13594   

   Lifetime Engaged Users  Lifetime Post Consumers  \
0                     178                      109   
1                    1457                     1361   
2                     177                      113   
3                    2211                      790   
4                     671                      410   

   Lifetime Post Consumptions  \
0                         159   
1                        1674   
2                         154   
3                        1119   
4                         580   

   Lifetime Post Impressions by people who have liked your Page  \
0                                               3078              
1                                              11710              
2                                               2812              
3                                              61027              
4                                               6228              

   Lifetime Post reach by people who like your Page  \
0                                              1640   
1                                              6112   
2                                              1503   
3                                             32048   
4                                              3200   

   Lifetime People who have liked your Page and engaged with your post  \
0                                                119                     
1                                               1108                     
2                                                132                     
3                                               1386                     
4                                                396                     

   comment  like  share  Total Interactions  
0        4    79     17                 100  
1        5   130     29                 164  
2        0    66     14                  80  
3       58  1572    147                1777  
4       19   325     49                 393

##Début de l'analyse

In [ ]:
# Calcul du taux d'engagement.
# Utilisation des colonnes 'Total Interactions' et 'Lifetime Post Total Reach' pour le calcul.
# Il est important de gérer les cas où 'Lifetime Post Total Reach' pourrait être égal à zéro pour éviter les erreurs de division.
df['Engagement Rate'] = np.where(
    df['Lifetime Post Total Reach'] > 0,
    ((df['Total Interactions'] / df['Lifetime Post Total Reach']) * 100).round(2),
    0
)

# Affichage et mise en pourcentage de la nouvelle colonne 'Engagement Rate'.
display(df[['Total Interactions', 'Lifetime Post Total Reach', 'Engagement Rate']].head().style.format({'Engagement Rate': '{:.2f}%'}))

In [ ]:
# 1. Vérifier s'il y a des valeurs Inf
print(f"Valeurs Inf : {df['Engagement Rate'].isin([np.inf, -np.inf]).sum()}")

# 2. Vérifier s'il y a des NaN
print(f"Valeurs NaN : {df['Engagement Rate'].isna().sum()}")

# 3. Vérifier le type de données
print(f"Type : {df['Engagement Rate'].dtype}")

# 4. Regarder quelques valeurs
print(df[['Total Interactions', 'Lifetime Post Total Reach', 'Engagement Rate']].head(20))

# 5. Stats descriptives
print(df['Engagement Rate'].describe())

Valeurs Inf : 0
Valeurs NaN : 0
Type : float64
    Total Interactions  Lifetime Post Total Reach  Engagement Rate
0                  100                       2752             3.63
1                  164                      10460             1.57
2                   80                       2413             3.32
3                 1777                      50128             3.54
4                  393                       7244             5.43
5                  186                      10472             1.78
6                  279                      11692             2.39
7                  339                      13720             2.47
8                  192                      11844             1.62
9                  142                       4694             3.03
10                 252                      21744             1.16
11                 106                       3112             3.41
12                 104                       2847             3.65
13             

Cette méthode nous donne un taux d'engagement direct, indiquant le pourcentage de personnes uniques ayant vu la publication et ayant interagi avec elle. C'est une métrique clé pour évaluer la performance des publications.

In [ ]:
# Répartition par type de post
print("Répartition des posts par type :\n")
display(df['Type'].value_counts())

Répartition des posts par type :



Type
Photo     426
Status     45
Link       22
Video       7
Name: count, dtype: int64

In [ ]:
# Taux d'engagement moyen par type de post
engagement_par_type = df.groupby('Type')['Engagement Rate'].mean().sort_values(ascending=False).round(2)

print("\nTaux d'engagement moyen par type de post :\n")
display(engagement_par_type)

# Type de post avec le taux d'engagement le plus élevé
type_plus_engageant = engagement_par_type.index[0]
taux_plus_engageant = engagement_par_type.iloc[0]

print(f"\nLe type de post avec le taux d'engagement le plus élevé est '{type_plus_engageant}' avec un taux moyen de {taux_plus_engageant:.2f}%.")


Taux d'engagement moyen par type de post :



Type
Photo     2.61
Status    1.59
Video     0.85
Link      0.73
Name: Engagement Rate, dtype: float64


Le type de post avec le taux d'engagement le plus élevé est 'Photo' avec un taux moyen de 2.61%.


In [ ]:
# Taux d'engagement moyen par heure de publication
engagement_par_heure = df.groupby('Post Hour')['Engagement Rate'].mean().sort_values(ascending=False).round(2)

print("\nTaux d'engagement moyen par heure de publication :\n")
display(engagement_par_heure)

# Heure de publication avec le taux d'engagement le plus élevé
heure_plus_engageante = engagement_par_heure.index[0]
taux_plus_engageant_heure = engagement_par_heure.iloc[0]

print(f"\nL'heure de publication qui génère le plus d'engagement est {heure_plus_engageante}h avec un taux moyen de {taux_plus_engageant_heure:.2f}%.")


Taux d'engagement moyen par heure de publication :



Post Hour
23    4.78
20    4.67
1     4.39
8     2.83
3     2.79
9     2.66
10    2.54
11    2.49
5     2.43
2     2.34
4     2.30
13    2.12
14    2.08
18    1.98
6     1.90
7     1.90
17    1.77
12    1.58
16    1.46
15    1.06
19    0.84
22    0.61
Name: Engagement Rate, dtype: float64


L'heure de publication qui génère le plus d'engagement est 23h avec un taux moyen de 4.78%.


In [ ]:
# Filtrer les posts de type 'Photo' publiés à 23h
posts_photo_23h = df[(df['Type'] == 'Photo') & (df['Post Hour'] == 23)]

# Compter le nombre de ces posts
nombre_posts_photo_23h = len(posts_photo_23h)

print(f"Nombre de posts de type 'Photo' publiés à 23h : {nombre_posts_photo_23h}")

Nombre de posts de type 'Photo' publiés à 23h : 1


Il y a ici un manque de pertinence avec le taux moyen le plus élevé à 23h, soutenu seulement par une publication. Il faut rappeler que le dataset contient 500 lignes uniques, ce qui est peu pour avoir des insights solides.

## Filtrer les posts Photo

Analyse plus poussée sur les posts de type 'Photo' qui représentent la majorité des publications et le taux d'engagement est le plus élevé.

In [ ]:
# Création d'un dataframe filtré sur les posts 'Photos'.
df_photos = df[df['Type'] == 'Photo']
display(df_photos.head())

Page total likes   Type  Post Month  Post Weekday  Post Hour  Paid  \
0            139441  Photo          12             4          3     0   
2            139441  Photo          12             3          3     0   
3            139441  Photo          12             2         10     1   
4            139441  Photo          12             2          3     0   
6            139441  Photo          12             1          3     1   

   Lifetime Post Total Reach  Lifetime Post Total Impressions  \
0                       2752                             5091   
2                       2413                             4373   
3                      50128                            87991   
4                       7244                            13594   
6                      11692                            19479   

   Lifetime Engaged Users  Lifetime Post Consumers  \
0                     178                      109   
2                     177                      113   
3                    2211                      790   
4                     671                      410   
6                     481                      265   

   Lifetime Post Consumptions  \
0                         159   
2                         154   
3                        1119   
4                         580   
6                         364   

   Lifetime Post Impressions by people who have liked your Page  \
0                                               3078              
2                                               2812              
3                                              61027              
4                                               6228              
6                                              15432              

   Lifetime Post reach by people who like your Page  \
0                                              1640   
2                                              1503   
3                                             32048   
4                                              3200   
6                                              9328   

   Lifetime People who have liked your Page and engaged with your post  \
0                                                119                     
2                                                132                     
3                                               1386                     
4                                                396                     
6                                                379                     

   comment  like  share  Total Interactions  Engagement Rate  
0        4    79     17                 100             3.63  
2        0    66     14                  80             3.32  
3       58  1572    147                1777             3.54  
4       19   325     49                 393             5.43  
6        3   249     27                 279             2.39

In [ ]:
# Calcul du taux d'engagement moyen pour les posts 'Photo' pour chaque jour de la semaine et identifier les jours avec le plus haut engagement.

engagement_photo_par_jour = df_photos.groupby('Post Weekday')['Engagement Rate'].mean().sort_values(ascending=False).round(2)

print("\nTaux d'engagement moyen des posts 'Photo' par jour de la semaine :\n")
display(engagement_photo_par_jour)

jour_plus_engageant_photo = engagement_photo_par_jour.index[0]
taux_plus_engageant_photo = engagement_photo_par_jour.iloc[0]

print(f"\nLe jour de la semaine avec le taux d'engagement le plus élevé pour les posts 'Photo' est le jour {jour_plus_engageant_photo} avec un taux moyen de {taux_plus_engageant_photo:.2f}%")


Taux d'engagement moyen des posts 'Photo' par jour de la semaine :



Post Weekday
5    2.97
1    2.78
4    2.68
2    2.64
6    2.54
7    2.48
3    2.17
Name: Engagement Rate, dtype: float64


Le jour de la semaine avec le taux d'engagement le plus élevé pour les posts 'Photo' est le jour 5 avec un taux moyen de 2.97%


In [ ]:
# Calcul du taux d'engagement moyen pour les posts 'Photo' pour chaque heure de la journée et identifier les heures avec le plus haut engagement.

engagement_photo_par_heure = df_photos.groupby('Post Hour')['Engagement Rate'].mean().sort_values(ascending=False).round(2)

print("\nTaux d'engagement moyen des posts 'Photo' par heure de publication :\n")
display(engagement_photo_par_heure)

heure_plus_engageante_photo = engagement_photo_par_heure.index[0]
taux_plus_engageant_photo_heure = engagement_photo_par_heure.iloc[0]

print(f"\nL'heure de publication qui génère le plus d'engagement pour les posts 'Photo' est {heure_plus_engageante_photo}h avec un taux moyen de {taux_plus_engageant_photo_heure:.2f}%.")


Taux d'engagement moyen des posts 'Photo' par heure de publication :



Post Hour
23    4.78
20    4.67
1     4.39
3     3.02
8     2.83
10    2.80
11    2.80
9     2.78
2     2.70
4     2.65
5     2.43
14    2.37
13    2.20
6     2.12
7     2.06
18    1.98
17    1.77
12    1.61
16    1.46
15    1.20
19    0.84
22    0.61
Name: Engagement Rate, dtype: float64


L'heure de publication qui génère le plus d'engagement pour les posts 'Photo' est 23h avec un taux moyen de 4.78%.


Même taux que précédement dont la pertinence est faible.
À prendre avec des pincettes.

## Insights supplémentaires

In [ ]:
# Impact des posts payants (Paid) sur le taux d'engagement
engagement_par_paid = df.groupby('Paid')['Engagement Rate'].mean().sort_values(ascending=False).round(2)
print("\nTaux d'engagement moyen pour les posts payants (sponsorisés) vs non-payants :\n")
display(engagement_par_paid)


taux_paid = engagement_par_paid.iloc[0]
print(f"\nLes posts non sponsorisés ont le taux d'engagement moyen le plus élevé ({taux_paid:.2f}%).")


Taux d'engagement moyen pour les posts payants (sponsorisés) vs non-payants :



Paid
0    2.48
1    2.23
Name: Engagement Rate, dtype: float64


Les posts où 'Paid' est 0 (non sponsorisé) ont le taux d'engagement moyen le plus élevé (2.48%). (0 = Non sponsorisé, 1 = Sponsorisé)


In [ ]:
# Calcul du taux d'engagement moyen par Mois de publication
engagement_par_mois = df.groupby('Post Month')['Engagement Rate'].mean().sort_values(ascending=False).round(2)
print("\nTaux d'engagement moyen par Mois de publication :\n")
display(engagement_par_mois)

mois_plus_engageant = engagement_par_mois.index[0]
taux_engageant_mois = engagement_par_mois.iloc[0]
print(f"\nLe mois de publication qui génère le plus d'engagement est le Mois {mois_plus_engageant} avec un taux moyen de {taux_engageant_mois:.2f}%.")


Taux d'engagement moyen par Mois de publication :



Post Month
9     4.17
5     3.00
11    2.83
8     2.65
12    2.43
10    2.43
7     2.36
4     2.04
1     2.00
6     1.95
2     1.79
3     1.13
Name: Engagement Rate, dtype: float64


Le mois de publication qui génère le plus d'engagement est le Mois 9 avec un taux moyen de 4.17%.


## Synthèse des Insights Clés

Voici un résumé des principales observations et recommandations de l'analyse du dataset :

1.  **Performance des Posts 'Photo' et Créneaux Optimaux :**
    *   Les **posts de type 'Photo'** sont le type de contenu le plus engageant en moyenne (taux d'engagement moyen de 2.61%).
    *   Pour les photos, le **Vendredi (Jour 5)** génère le taux d'engagement moyen le plus élevé (2.97%).
    *   L'**heure de publication optimale est 23h**, avec un taux d'engagement moyen de 4.78% pour les photos.
    *   *Insight :* Privilégier la publication de photos le Vendredi soir pour maximiser l'engagement.

2.  **Impact des Posts Sponsorisés (Paid) :**
    *   Les posts **non sponsorisés (Paid = 0)** montrent un taux d'engagement moyen légèrement supérieur (2.48%) aux posts sponsorisés (Paid = 1) qui sont à 2.23%.
    *   *Insight :* Il est intéressant de noter que l'engagement organique peut être plus fort, ou que les campagnes payantes attirent une audience plus large mais potentiellement moins réceptive.

3.  **Taux d'Engagement par Mois :**
    *   Le **Mois de Septembre (Mois 9)** se distingue comme le mois le plus performant en termes d'engagement moyen, avec un taux de 4.17%.
    *   *Insight :* Considérer une attention particulière et peut-être une augmentation de l'activité de publication en Septembre. (Impossible de savoir si c'est saisonnier car le dataset est sur une durée d'un an)

In [ ]:
# Enregistrement du dataframe nettoyé et avec tous les types dans le dossier drive
nouveau_nom_fichier = 'dataset_Facebook_cleaned.csv'
chemin_nouveau_fichier = os.path.join(dossier, nouveau_nom_fichier)

df.to_csv(chemin_nouveau_fichier, index=False, sep=';', decimal=',', encoding='utf-8')

print(f"Le DataFrame a été sauvegardé avec succès sous : {chemin_nouveau_fichier}")

Le DataFrame a été sauvegardé avec succès sous : /content/drive/My Drive/Colab Notebooks/Projet_Facebook_Dataset/dataset_Facebook_cleaned.csv
